client/relay-chain-minimal-node/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use collator_overseer::{CollatorOverseerGenArgs, NewMinimalNode};
+
+use cumulus_relay_chain_interface::RelayChainError;
+use polkadot_network_bridge::{peer_sets_info, IsAuthority};
+use polkadot_node_network_protocol::{
+	peer_set::PeerSetProtocolNames,
+	request_response::{self, IncomingRequest, ReqProtocolNames},
+};
+use polkadot_node_subsystem_util::metrics::prometheus::Registry;
+use polkadot_primitives::v2::CollatorPair;
+
+use sc_authority_discovery::Service as AuthorityDiscoveryService;
+use sc_network::{Event, NetworkService};
+use sc_network_common::service::NetworkEventStream;
+use std::sync::Arc;
+
+use polkadot_service::{open_database, Configuration, TaskManager};
+
+use futures::StreamExt;
+
+use sp_runtime::traits::Block as BlockT;
+
+mod collator_overseer;
+
+mod network;
+
+mod blockchain_rpc_client;
+pub use blockchain_rpc_client::BlockChainRpcClient;
+
+fn build_authority_discovery_service<Block: BlockT>(
+	task_manager: &TaskManager,
+	client: Arc<BlockChainRpcClient>,
+	config: &Configuration,
+	network: Arc<NetworkService<Block, <Block as BlockT>::Hash>>,
+	prometheus_registry: Option<Registry>,
+) -> AuthorityDiscoveryService {
+	let auth_disc_publish_non_global_ips = config.network.allow_non_globals_in_dht;
+	let authority_discovery_role = sc_authority_discovery::Role::Discover;
+	let dht_event_stream = network.event_stream("authority-discovery").filter_map(|e| async move {
+		match e {
+			Event::Dht(e) => Some(e),
+			_ => None,
+		}
+	});
+	let (worker, service) = sc_authority_discovery::new_worker_and_service_with_config(
+		sc_authority_discovery::WorkerConfig {
+			publish_non_global_ips: auth_disc_publish_non_global_ips,
+			// Require that authority discovery records are signed.
+			strict_record_validation: true,
+			..Default::default()
+		},
+		client,
+		network.clone(),
+		Box::pin(dht_event_stream),
+		authority_discovery_role,
+		prometheus_registry.clone(),
+	);
+
+	task_manager.spawn_handle().spawn(
+		"authority-discovery-worker",
+		Some("authority-discovery"),
+		worker.run(),
+	);
+	service
+}
+
+/// Builds a minimal relay chain node. Chain data is fetched
+/// via [`BlockChainRpcClient`] and fed into the overseer and its subsystems.
+///
+/// Instead of spawning all subsystems, this minimal node will only spawn subsystems
+/// required to collate:
+/// - AvailabilityRecovery
+/// - CollationGeneration
+/// - CollatorProtocol
+/// - NetworkBridgeRx
+/// - NetworkBridgeTx
+/// - RuntimeApi
+#[sc_tracing::logging::prefix_logs_with("Relaychain")]
+pub async fn new_minimal_relay_chain(
+	mut config: Configuration,
+	collator_pair: CollatorPair,
+	relay_chain_rpc_client: Arc<BlockChainRpcClient>,
+) -> Result<NewMinimalNode, RelayChainError> {
+	let role = config.role.clone();
+
+	// Use the given RPC node as bootnode, since we do not have a chain spec with valid boot nodes
```

Comment:
I'm okay with this. However, we should probably somewhere up in the stack returning an error if someone passes arguments to the relay chain node, which we consume, but ignore.

Reply:
Since I am passing the config into the polkadot code, it is a bit hard to analyze which fields exactly are used and which are unused. I added a warning message to the cli that prints a message when users provide `--relay-chain-rpc-url` together with raw relay chain arguments.